src/transformers/models/propainter/modeling_propainter.py

```diff
@@ -0,0 +1,4583 @@
+            # end for
+            if "backward" in module_name:
+                features[module_name] = features[module_name][::-1]
+                masks[module_name] = masks[module_name][::-1]
+
+        outputs_b = torch.stack(features["backward_1"], dim=1).view(-1, num_channels, height, width)
+        outputs_f = torch.stack(features["forward_1"], dim=1).view(-1, num_channels, height, width)
+
+        if self.learnable:
+            mask_in = mask.view(-1, 2, height, width)
+            masks_f = None
+            outputs = self.fuse(torch.cat([outputs_b, outputs_f, mask_in], dim=1)) + masked_frames.view(
+                -1, num_channels, height, width
+            )
+        else:
+            masks_f = torch.stack(masks["forward_1"], dim=1)
+            outputs = outputs_f
+
+        return (
+            outputs_b.view(batch_size, -1, num_channels, height, width),
+            outputs_f.view(batch_size, -1, num_channels, height, width),
+            outputs.view(batch_size, -1, num_channels, height, width),
+            masks_f,
+        )
+
+
+class ProPainterDeconv(nn.Module):
+    def __init__(
+        self,
+        input_channel: int,
+        output_channel: int,
+        kernel_size: int = 3,
+        padding: int = 0,
+    ):
+        super().__init__()
+        self.conv = nn.Conv2d(
+            input_channel,
+            output_channel,
+            kernel_size=kernel_size,
+            stride=1,
+            padding=padding,
+        )
+
+    def forward(self, hidden_states):
+        hidden_states = F.interpolate(hidden_states, scale_factor=2, mode="bilinear", align_corners=True)
+        return self.conv(hidden_states)
+
+
+class ProPainterModulatedDeformConv2d(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: int,
+        stride: int = 1,
+        padding: int = 0,
+        dilation: int = 1,
+        groups: int = 1,
+        deform_groups: int = 1,
+        bias: bool = True,
+    ):
+        super().__init__()
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.kernel_size = _pair(kernel_size)
+        self.stride = stride
+        self.padding = padding
+        self.dilation = dilation
+        self.groups = groups
+        self.deform_groups = deform_groups
+        self.with_bias = bias
+        # enable compatibility with nn.Conv2d
+        self.transposed = False
+        self.output_padding = _single(0)
+
+        self.weight = nn.Parameter(torch.Tensor(out_channels, in_channels // groups, *self.kernel_size))
+        if bias:
+            self.bias = nn.Parameter(torch.Tensor(out_channels))
+        else:
+            self.register_parameter("bias", None)
+
+    def forward(self, hidden_states, offset, mask):
+        pass
+
+
+class ProPainterDeformableAlignment(ProPainterModulatedDeformConv2d):
+    """Second-order deformable alignment module."""
```

Comment:
this is first-order, no?

Reply:
in the original code it's mentioned as second order module which is also followed here but the core operation and design principles align more closely with first order deformable convolutions due to the way offsets are generated and applied. The term `second order` in this context might refer to additional features that go beyond standard first order deformable module